File: providers/perplexity/human.js
// Human-behaviour helpers for Playwright interactions.
//
// Automation detection (Cloudflare Turnstile, Perplexity's own gates) looks for
// robotic patterns: instant teleport clicks, uniform keystroke intervals, zero
// idle time. These helpers add realistic jitter so interactions look like a
// person rather than a script.

export function randomDelay(minMs, maxMs) {
  return Math.floor(Math.random() * (maxMs - minMs + 1)) + minMs;
}

export async function humanPause(page, minMs, maxMs) {
  await page.waitForTimeout(randomDelay(minMs, maxMs));
}

// Typo characters near common letters on QWERTY — cheap approximation of
// fat-finger mistakes. Not cryptographically real but looks plausible.
const NEAR_KEY = {
  a: 's', s: 'd', d: 'f', f: 'g', g: 'h', h: 'j', j: 'k', k: 'l',
  q: 'w', w: 'e', e: 'r', r: 't', t: 'y', y: 'u', u: 'i', i: 'o', o: 'p',
  z: 'x', x: 'c', c: 'v', v: 'b', b: 'n', n: 'm',
};

// Types `text` character-by-character via `page.keyboard`, with per-char delay
// jitter, occasional pauses, and a small rate of typo-then-backspace
// corrections. Assumes focus is already on the target input.
export async function humanType(page, text, { typoChance = 0.025 } = {}) {
  for (const char of text) {
    const lower = char.toLowerCase();
    const nearby = NEAR_KEY[lower];
    if (nearby && Math.random() < typoChance) {
      // Type a wrong char, pause like noticing the typo, backspace, carry on
      const wrong = char === char.toUpperCase() ? nearby.toUpperCase() : nearby;
      await page.keyboard.type(wrong, { delay: randomDelay(45, 130) });
      await page.waitForTimeout(randomDelay(160, 420));
      await page.keyboard.press('Backspace');
      await page.waitForTimeout(randomDelay(80, 200));
    }
    await page.keyboard.type(char, { delay: randomDelay(45, 130) });
    if (Math.random() < 0.08) {
      await page.waitForTimeout(randomDelay(120, 320));
    }
  }
}

// Moves the mouse to the locator's box with a few intermediate waypoints
// (Playwright's default `.click()` teleports; Bezier-ish paths are closer to
// real input), hovers briefly, then clicks with a non-zero mousedown→up delay.
export async function humanClick(page, locator, { timeout = 10_000 } = {}) {
  const handle = await locator.first().elementHandle({ timeout });
  if (!handle) throw new Error('humanClick: element not found');
  const box = await handle.boundingBox();
  if (!box) {
    // Element has no box (display:none or off-screen) — fall back to .click()
    await locator.first().click({ timeout, delay: randomDelay(40, 110) });
    return;
  }

  const targetX = box.x + box.width / 2 + randomDelay(-Math.max(1, box.width / 8), Math.max(1, box.width / 8));
  const targetY = box.y + box.height / 2 + randomDelay(-Math.max(1, box.height / 8), Math.max(1, box.height / 8));

  // Move through 2-3 intermediate points for a non-linear path
  const steps = 2 + Math.floor(Math.random() * 2);
  const current = await page.evaluate(() => ({ x: window.innerWidth / 2, y: window.innerHeight / 2 }));
  for (let i = 1; i <= steps; i++) {
    const t = i / (steps + 1);
    const jitterX = randomDelay(-12, 12);
    const jitterY = randomDelay(-12, 12);
    const x = current.x + (targetX - current.x) * t + jitterX;
    const y = current.y + (targetY - current.y) * t + jitterY;
    await page.mouse.move(x, y, { steps: 4 });
    await page.waitForTimeout(randomDelay(20, 80));
  }
  await page.mouse.move(targetX, targetY, { steps: 6 });
  await page.waitForTimeout(randomDelay(100, 280));
  await page.mouse.down();
  await page.waitForTimeout(randomDelay(40, 110));
  await page.mouse.up();
}

File: ask-perplexity-cli/src/automation/human.rs
use anyhow::Result;
use chromiumoxide::Element;
use rand::{rng, RngExt};
use std::time::Duration;
use tokio::time::sleep;

/// Sleep a random duration between `min_ms` and `max_ms` (inclusive).
pub async fn human_pause(min_ms: u64, max_ms: u64) {
    let delay = if min_ms >= max_ms {
        min_ms
    } else {
        rng().random_range(min_ms..=max_ms)
    };
    sleep(Duration::from_millis(delay)).await;
}

/// Type `text` into `element` with per-char delay jitter, occasional
/// pauses, and a small rate of typo-then-backspace corrections.
/// Mirrors providers/perplexity/human.js#humanType from Plan 1.
pub async fn human_type(element: &Element, text: &str) -> Result<()> {
    const NEARBY: &[(&str, &str)] = &[
        ("a", "s"), ("s", "d"), ("d", "f"), ("f", "g"), ("g", "h"),
        ("h", "j"), ("j", "k"), ("k", "l"), ("q", "w"), ("w", "e"),
        ("e", "r"), ("r", "t"), ("t", "y"), ("y", "u"), ("u", "i"),
        ("i", "o"), ("o", "p"), ("z", "x"), ("x", "c"), ("c", "v"),
        ("v", "b"), ("b", "n"), ("n", "m"),
    ];
    let find_nearby = |c: char| -> Option<char> {
        let lower = c.to_ascii_lowercase().to_string();
        NEARBY
            .iter()
            .find(|(k, _)| *k == lower)
            .map(|(_, v)| {
                if c.is_ascii_uppercase() {
                    v.chars().next().unwrap().to_ascii_uppercase()
                } else {
                    v.chars().next().unwrap()
                }
            })
    };

    for ch in text.chars() {
        let ch = match ch {
            '\n' | '\r' | '\t' => ' ',
            _ => ch,
        };
        let typo_roll: u32 = rng().random_range(0..1000);
        if typo_roll < 25 {
            if let Some(wrong) = find_nearby(ch) {
                element.type_str(&wrong.to_string()).await?;
                human_pause(160, 420).await;
                element.press_key("Backspace").await?;
                human_pause(80, 200).await;
            }
        }
        let delay = rng().random_range(45..=130);
        element.type_str(&ch.to_string()).await?;
        sleep(Duration::from_millis(delay)).await;
        if rng().random_range(0..100) < 8 {
            human_pause(120, 320).await;
        }
    }
    Ok(())
}

/// Click an element with realistic mouse trajectory + a non-zero
/// down/up delay. chromiumoxide's default .click() teleports.
pub async fn human_click(page: &chromiumoxide::Page, element: &Element) -> Result<()> {
    // chromiumoxide 0.7 doesn't expose a simple `move_mouse` — use CDP directly.
    // bounding_box() returns Result<BoundingBox> directly in 0.7.
    let bb = element.bounding_box().await?;
    let target_x = bb.x + bb.width / 2.0;
    let target_y = bb.y + bb.height / 2.0;

    // A couple of intermediate waypoints with jitter
    let (vw, vh) = (1440.0, 900.0);
    let current_x = vw / 2.0;
    let current_y = vh / 2.0;
    let steps = 3u32;
    for i in 1..=steps {
        let t = i as f64 / (steps + 1) as f64;
        let jx: f64 = rng().random_range(-12..=12) as f64;
        let jy: f64 = rng().random_range(-12..=12) as f64;
        let x = current_x + (target_x - current_x) * t + jx;
        let y = current_y + (target_y - current_y) * t + jy;
        page.execute(
            chromiumoxide::cdp::browser_protocol::input::DispatchMouseEventParams::builder()
                .r#type(chromiumoxide::cdp::browser_protocol::input::DispatchMouseEventType::MouseMoved)
                .x(x)
                .y(y)
                .build()
                .unwrap(),
        )
        .await?;
        human_pause(20, 80).await;
    }

    // Down-wait-up for a realistic click duration
    page.execute(
        chromiumoxide::cdp::browser_protocol::input::DispatchMouseEventParams::builder()
            .r#type(chromiumoxide::cdp::browser_protocol::input::DispatchMouseEventType::MousePressed)
            .x(target_x)
            .y(target_y)
            .button(chromiumoxide::cdp::browser_protocol::input::MouseButton::Left)
            .click_count(1)
            .build()
            .unwrap(),
    )
    .await?;
    human_pause(40, 110).await;
    page.execute(
        chromiumoxide::cdp::browser_protocol::input::DispatchMouseEventParams::builder()
            .r#type(chromiumoxide::cdp::browser_protocol::input::DispatchMouseEventType::MouseReleased)
            .x(target_x)
            .y(target_y)
            .button(chromiumoxide::cdp::browser_protocol::input::MouseButton::Left)
            .click_count(1)
            .build()
            .unwrap(),
    )
    .await?;
    Ok(())
}
